### PR TITLE
Add support for the tagfunc option

### PR DESCRIPTION
**Pull request by johnschug**
Adds a new function that can be used with the `tagfunc` option added in vim patch 8.1.1228 (neovim/neovim#11199).

Also consolidates various synchronous wait implementations in favor of a polyfill based on the neovim `wait` function.

### CODE REVIEW

**Comment by prabirshrestha** [comment]
can u also add docs?

**Comment by johnschug** [comment]
Added docs and rebased on master

File: autoload/lsp/tag.vim [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        let l:range = a:loc['range']
+    endif
+
+    if !lsp#utils#is_file_uri(l:uri)
+        return v:null
+    endif
+
+    return {
+        \ 'filename': lsp#utils#uri_to_path(l:uri),
+        \ 'cmd': string(l:range['start']['line'] + 1)
```
Comment (prabirshrestha) [inline autoload/lsp/tag.vim:22]: Can you instead use functions from https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/utils/position.vim so it works for multi byte chars.

File: autoload/lsp/tag.vim [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+function! s:handle_locations(ctx, server, type, data) abort
+    try
+        if lsp#client#is_error(a:data['response']) || !has_key(a:data['response'], 'result')
+            call lsp#utils#error('Failed to retrieve ' . a:type . ' for ' . a:server . ': ' . lsp#client#error_message(a:data['response']))
+            return
+        endif
+
+        let l:list = a:ctx['list']
+        let l:result = a:data['response']['result']
+        if type(l:result) != v:t_list
```
Comment (prabirshrestha) [inline autoload/lsp/tag.vim:35]: `v:t_list` might not be supported in older version of vim/neovim. use `type([])` instead.

File: autoload/lsp/ui/vim/utils.vim [hunk 3]
```diff
@@ -135,7 +135,7 @@ function! lsp#ui#vim#utils#diagnostics_to_loc_list(result) abort
     return l:list
 endfunction
 
-function! s:get_symbol_text_from_kind(server, kind) abort
+function! lsp#ui#vim#utils#get_symbol_text_from_kind(server, kind) abort
```
Comment (prabirshrestha) [inline autoload/lsp/ui/vim/utils.vim:138]: Can we pefix `get_symbol_text_from_kind` with `_`. I would like to make this method internal for now.

**Comment by prabirshrestha** [comment]
Also appreciate the extra refactor you added to this PR.

**Comment by johnschug** [comment]
Fixed

**Comment by prabirshrestha** [comment]
Looks good. Merged. Thanks.

Also one good option would be to also echom the messages such as retrieving tag or tag retreived since some language servers are slow.

**Pull request merged**